stream_alert/shared/publisher.py

```diff
@@ -0,0 +1,272 @@
+            thing (mixed): Any primitive or reference to be checked
+
+        Returns:
+            bool
+        """
+
+        # We have to put the isclass() check BEFORE the callable() check because classes are also
+        # callable!
+        if isclass(thing) and issubclass(thing, AlertPublisher):
+            return True
+        elif callable(thing):
+            return True
+
+        return False
+
+    @staticmethod
+    def get_publisher_name(class_or_function):
+        """Given a class or function, will return its fully qualified name.
+
+        This is useful for assigning a unique string name for a publisher.
+
+        Args:
+            class_or_function (callable|Class): A reference to a python function or class
+
+        Returns:
+            string
+        """
+        return '{}.{}'.format(class_or_function.__module__, class_or_function.__name__)
+
+    @classmethod
+    def register_publisher(cls, publisher):
+        """Registers the publisher into the repository.
+
+        To standardize the interface of publishers, if a function publisher is given, it will be
+        wrapped with a WrappedFunctionPublisher instance prior to being registed into the
+        Repository.
+
+        Args:
+             publisher (callable|AlertPublisher): An instance of a publisher class or a function
+        """
+        if not AlertPublisherRepository.is_valid_publisher(publisher):
+            LOGGER.error(
+                'Could not register publisher %s; Not callable nor subclass of AlertPublisher',
+                publisher
+            )
+            raise PublisherRegistrationError
+
+        elif isclass(publisher):
+            # If the provided publisher is a Class, then we simply need to instantiate an instance
+            # of the class and register it.
+            publisher_instance = publisher()
+        else:
+            # If the provided publisher is a function, we wrap it with a WrappedFunctionPublisher
+            # to make them easier to handle.
+            publisher_instance = WrappedFunctionPublisher(publisher)
+
+        name = AlertPublisherRepository.get_publisher_name(publisher)
+
+        if name in cls._publishers:
+            LOGGER.error('Publisher with name [%s] has already been registered.', name)
+            raise PublisherRegistrationError
+
+        cls._publishers[name] = publisher_instance
+
+    @classmethod
+    def get_publisher(cls, name):
+        """Returns the publisher with the given name
+
+        Args:
+            name (str): The name of the publisher.
+
+        Returns:
+            AlertPublisher|None
+        """
+        if cls.has_publisher(name):
+            return cls._publishers[name]
+
+        LOGGER.error('Publisher [%s] does not exist', name)
+
+    @classmethod
+    def has_publisher(cls, name):
+        """Returns true if the given publisher name has been registered in this Repository
+        """
+        AlertPublisherImporter.import_publishers()
```

Comment:
it still looks like this could be called a bunch of times.. could this call just be put at the class level (since it doesn't appear that this class is ever instantiated) so it's only called once at import time? like:

```python
class AlertPublisherRepository(object):
    """A repository mapping names -> publishers
     As a usability optimization, using this Repository will eagerly load and register all
    publishers in the application.
    """
    _publishers = {}

    AlertPublisherImporter.import_publishers()  # do import now, once, at import time
```

TBH I don't know if you need a dedicated `AlertPublisherImporter` class since it mostly just seems to be a passthrough to the `import_folders` function anyway. Thoughts?

Reply:
wtf you can put code that executes in a class outside of a function :mindblown:

Reply:
Nope.. just tried... can't do this because it causes a cyclical require dependency in the following:

`stream_alert/shared/publisher` -> `publishers/community/generic` -> `stream_alert.shared.publisher`

I could probably do this if I split it into its own file, because the `publishers/community/generic` requires the `@Register` annotation, not the repository itself. Will think about it.